docs/decisions/0013-library-reference-content-block.rst

```diff
@@ -0,0 +1,61 @@
+
+We are currently building out the functionality of libraries v2, and trying to migrate all usage of content libraries onto Blockstore.
+The benefits of using the Blockstore backed v2 libraries are:
+
+#. Blockstore is much more flexible than modulestore and can store anything that can be represented as a file - unlike modulestore, which only stores courses built of XBlocks. Also, the Blockstore is being moved into the edx-platform.
+#. simpler to maintain than modulestore
+#. increases re-usability of the content
+
+Terminology Used
+^^^^^^^^^^^^^^^^
+#. Blocks: Each reusable piece of content in a library. Example: problems, videos, HTML. Also known as Xblocks.
+#. Library: Collection of blocks in a course.
+
+
+Decisions
+=========
+We want to be able to reference single and multiple specific blocks from a library (v2, as well as v1) into a course (that are stored in modulestore)
+The workflow discussed for the interface for library content referencing would be:
+
+#. Author chooses a library.
+#. Author chooses a pool of blocks.
+#. If multiple blocks of the same type are selected, ask for randomization
+#. If no randomization, then ask for ordering
+
+Current Architecture/Implementation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Currently, the courses are stored in modulestore and the libraries can either be stored in modulestore or blockstore.
+Since the course lives in the modulestore and some of the libraries in blockstore, we use the [library_sourced_block](https://github.com/openedx/edx-platform/blob/master/common/lib/xmodule/xmodule/library_sourced_block.py) to make a copy of that blockstore based library block and store it in modulestore itself as the child.
+
+
+Proposed Implementation
+^^^^^^^^^^^^^^
+The library referencing flow would be implemented within the Studio and/or LMS processes of a single Open edX instance.
+The library content would be stored and rendered outside the modulestore.
+
+We would achieve this by introducing a unit_compositor subsystem. Like learning_sequences, the subsystem would be populated by CMS upon course publish. It would store a read-optimized form of:
+* Metadata and child-block lists for all course Units, and
+* Definitions of library blocks.
+
+We would then update the LmsXBlockRuntime (called “CombinedSystem” until BD-13 is done) to use the unit_compositor as its backing store for units. When a unit is requested for a particular user, the unit_compositor would:
+#. Load the unit’s child blocks from modulestore.
+#. Replace each library reference block with its corresponding library block definitions, each overridden with any course-author-specified customizations, and each given a usage key that composes the library reference block's usage information with the library block’s definition key.
+#. Return the list of blocks wrapped under a VerticalBlock, with the same usage key as the original unit, for the LmsXBlockRuntime to render.
+
```

Comment:
Is it feasible to make this unit compositor a general mechanism for UserPartitions, instead of being specific to content libraries?

Reply:
@ormsbee I think that was the idea but in this ADR we are strictly sticking to content libraries.

Maybe @marcotuts or @jristau1984 can shed more light here.

Reply:
@saksham115, @jristau1984: I'd like to explore what a more general model could look like, and what a feasible migration path could be. For instance, do we need to access the modulestore before doing the substitutions? If there's an independent data model powering the compositor, could the composition happen before the modulestore grabs the data? Would that give us a smoother path to shifting more things (e.g. split tests, enrollment track partitions) out of modulestore/XBlock logic and into the compositor over time?

Reply:
Agreed with @ormsbee - if this unit_compositor is going to be for units what learning_sequences is for sections/subsections - an optimized, non-modulestore-based data source - then it should ideally be able to support all the functional pieces that might serve different units to different learners without using the modulestore to do so. It's worth putting some thought into - and possibly addressing in this ADR.

Reply:
@ormsbee For now I see we have a dependency on modulestore to store the reference for the blocks. We do need to answer questions like, how do we know what blocks are placed in the course without accessing Modulestore?

Since courses are still stored in Modulestore.

Although the way I see unit compositor, it's agnostic of Modulestore what it should need is some input like [this example](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3173220481/BD-14+Library+use+cases+implementation+discovery#Example%3A) and it transforms it into a Vertical Block with all the reference expanded.

I can probably be overlooking some complexity so please feel free to correct me if I am wrong.

cc: @bradenmacdonald

Reply:
@ormsbee  - can we get an update here on latest thinking on what the general approach for the unit compositor looks like? I'd like to be able to deliver BD-14's benefits without having to delay for a general solution, but I'm not sure how the thinking has evolved.

Reply:
I think we can still build out the improved library authoring workflows and the "use content from library in course" workflows, and just have them still do a copy from blockstore to modulestore for now. Then whenever the unit compositor is ready we can change the implementation to use that, which should have nice performance improvements but not necessarily any changes to the UX.

Reply:
Sorry I'm late to reply on this. I agree with @bradenmacdonald's assessment here.